File: Java/012-Highly-divisible-triangular-number.java
//It is based on coprime numbers properties. The triangle number is equal to n*(n+1)/2 where n and n+1 are coprimes => n and (n+1)/2 are coprimes 
//or n/2 and n+1 are coprimes ( it depends if n is even or odd )
//I used an integer as argument for getTriangle() method to be easier to improve this method. 
//You can decide to start from a value that is bigger than 1 for this particular case. 
//In this case, you can reduce the time at least 3 times

public class P12 {

    final static int THRESHOLD_DIVISORS = 500;    

    public static void main(String[] args) {
        System.out.println(getTriangle(1));
    }

    public static long getTriangle(int n) {
        long numberOfDivisors = 0;
        long firstCoprime, secondCoprime;
        while (true) {
            if ( n % 2 == 0 ) {
                firstCoprime = getNumberOfDivisors(n/2);
                secondCoprime = getNumberOfDivisors(n+1);

            } else {
                firstCoprime = getNumberOfDivisors(n);
                secondCoprime = getNumberOfDivisors((n+1)/2);
            }           
            numberOfDivisors = firstCoprime * secondCoprime;
            if ( numberOfDivisors > THRESHOLD_DIVISORS ) {
                return n*(n+1)/2;
            }       
            n++;
        }
    }

    private static long getNumberOfDivisors(long nr) {
        int divisors = 0;
        int i;
        int sqrt = (int) Math.sqrt(nr);
        for ( i = 1 ; i <= sqrt ; i++ ) {
            if ( nr % i == 0 ) {
                divisors+=2;           // E.g.: (2, n/2), (3, n/3)
            }
        }
        if ( sqrt*sqrt == nr ){        // it was counted twice
            divisors--;
        }
        return divisors;
    }

}

File: C++/012-Highly_divisible_triangular_number.cpp
#include <iostream>
#include <map>
#include <math.h>

int main()
{
	int tn;
	int c = 1;
	while (true)
	{
		c++;
		tn = c*(c+1)/2;
		int n = tn;
		// find prime factors of number
		
		std::map<std::string, int> factor_map;
		while (n % 2 == 0)  
    		{
			if (factor_map.find("2") == factor_map.end())
				factor_map["2"] = 1;
			else
				factor_map["2"] = factor_map["2"] + 1;
        		n = n/2;
    		}

		for (int i = 3; i <= sqrt(n); i = i + 2)
		{
			while (n % i == 0)
			{
				if (factor_map.find(std::to_string(i)) == factor_map.end())
                                	factor_map[std::to_string(i)] = 1;
                        	else
                                	factor_map[std::to_string(i)] = factor_map[std::to_string(i)] + 1;
				n = n/i;
			}
		}

		if (n > 2)
		{
			if (factor_map.find(std::to_string(n)) == factor_map.end())
                                factor_map[std::to_string(n)] = 1;
                        else
                        	factor_map[std::to_string(n)] = factor_map[std::to_string(n)] + 1;
		}

		int divisors = 1;
		std::map<std::string, int>::iterator it;
		for (it = factor_map.begin(); it != factor_map.end(); it++)
		{
			divisors *= it->second + 1;
		}

		if (divisors > 500)
			break;
	}
	std::cout << "found: " << tn << std::endl;
}
